### DISCUSSION TITLE
Would it make sense to use Remix to render a headless CMS?

### DISCUSSION THREAD

**Discussion by AndrewLeedham** [remix-run/remix#953] [Q&A] [answered]
From what I can tell Remix relies heavily on knowing what routes exist so it can optimise what is loading. However, for a headless CMS you would not necessarily know said routes up front as they are stored behind an API call. In Express for example this would likely be implemented as a catch-all route at the end of the named ones (`app.use('*', cmsHandler)` for example).

So I guess the question is:
Say we are using Remix with the Express adapter, would it make sense to move `app.use('*', cmsHandler)` into Remix or keep it in Express land? (This is also under the assumption that the CMS gives back structured data which can be fed into React components, so it would be using Remix's rendering if it was moved to Remix).

**Comment by sheepsteak** [comment, marked as answer]
Wouldn't you just use a [splat route](https://remix.run/docs/en/v1/api/conventions#splat-routes) and then in the `loader` work out if that is a valid URL or not? If it is you'd get the data and return it and then the React component works out what to render from it. If it's not a valid route then [return a 404](https://remix.run/docs/en/v1/guides/not-found#how-to-send-a-404).

**Reply by AndrewLeedham** [reply]
I hadn't come across splat routes. At least not the named concept I was aware of the `$` syntax.

It is possible to use Remix in this way, I was more wondering if it makes sense. That is does it negate the advantages of Remix (Nested Routing). I suppose you could still have common UI like header/footer in a layout/app component so they are retained between navigations.

**Comment by ccssmnn** [comment]
Is the structure of the pages arbitrary? Assuming it is then using remix, you still get Rendering and preloading Links. I would say it still makes sense.

**Reply by AndrewLeedham** [reply]
> Is the structure of the pages arbitrary?
Yes page URLs are defined in the CMS. And can be anything.

I agree Remix has so much niceness for just using React as a server-side templating languages, even without nested routing!

**Reply by ccssmnn** [reply]
Remix gives you the tools to build whatever website or app you want. It's so damn straightforward. I've built a Headless WordPress site and currently developing an app with Firebase. In both cases, remixes non-abstractions are performant and clean.
